realis.go

```diff
@@ -428,15 +418,51 @@ func basicAuth(username, password string) string {
 	return base64.StdEncoding.EncodeToString([]byte(auth))
 }
 
+type auroraThriftCall func() (resp *aurora.Response, err error)
+
+// Takes a Thrift API function call and returns response and error.
+// If Error from the API call is retryable, the functions re-establishes the connection with Aurora by
+// using the same configuration used by the original client. Locks usage of and changes to client connection in order
+// to make realis sessions thread safe.
+func (r *realisClient) thriftCallHelper(auroraCall auroraThriftCall) (*aurora.Response, error) {
+	// Only allow one go-routine make use or modify the thrift client connection
+	r.lock.Lock()
```

Comment:
Use defer to Unlock?

Reply:
If we do that then we unlock right before the function returns. I wanted to make the critical section as short as possible while still avoiding race conditions.

Reply:
defer guarantees execution no matter how or from where it runs.. panic scenario or not.. besides, rest of the logic after unlock is pretty light -- interpretation of response code.

Reply:
It does allow someone using the library to recover with the lock released in that case. Not sure how much that really buys us but I'm fine with it as long as the size of this function doesn't increase too much down the line.